fix(compiler-cli): prefer non-aliased exports in reference emitters

JoostK opened the pull request:
This commit changes the reference emitters in the Ivy compiler to prefer
non-aliased exports if they exist. This avoids selecting "private
exports" that may not be stable, e.g. the reexports that have been added
by the View Engine compiler. Such reexports are not stable and are
therefore not suitable to be emitted into partial compilations, as the
output of partial compilations should only reference stable symbols
from upstream libraries.

An alternative solution has been considered where ViewEngine-generated
exports would gain a certain prefix, such that the Ivy compiler could
just exclude those exports (see #41443). However, that solution would
be insufficient in case a library is built using partial compilation and
using a VE-compiled library from earlier versions of Angular, where the
magic prefix would be missing. For such libraries, ngcc would have
generated reexports using the declared name if not already present so
this change does result in choosing the correct export.

Because ngcc always generates reexports using the declared name even if
an aliased export is present, this change causes those ngcc-generated
exports to be chosen in downstream libraries using partial compilation.
This is unfortunate as it means that the declared names become
effectively public even if the library author was intentionally
exporting it using an alias. This commit does not address this problem;
it is expected that this should not result in widespread issues across
the library ecosystem.

Fixes #41277

JoostK committed:
fix(compiler-cli): prefer non-aliased exports in reference emitters

This commit changes the reference emitters in the Ivy compiler to prefer
non-aliased exports if they exist. This avoids selecting "private
exports" that may not be stable, e.g. the reexports that have been added
by the View Engine compiler. Such reexports are not stable and are
therefore not suitable to be emitted into partial compilations, as the
output of partial compilations should only reference stable symbols
from upstream libraries.

An alternative solution has been considered where ViewEngine-generated
exports would gain a certain prefix, such that the Ivy compiler could
just exclude those exports (see #41443). However, that solution would
be insufficient in case a library is built using partial compilation and
while depending itself on a VE-compiled library from earlier versions of
Angular, where the magic prefix would be missing. For such libraries,
ngcc would have generated reexports using the declared name if not already
present so this change does result in choosing the correct export.

Because ngcc always generates reexports using the declared name even if
an aliased export is present, this change causes those ngcc-generated
exports to be chosen in downstream libraries using partial compilation.
This is unfortunate as it means that the declared names become
effectively public even if the library author was intentionally
exporting it using an alias. This commit does not address this problem;
it is expected that this should not result in widespread issues across
the library ecosystem.

Fixes #41277

petebacondarwin commented:
Could you consider rewording this sentence from the commit message:

> However, that solution would be insufficient in case a library is built using partial compilation **and using** a VE-compiled library from earlier versions of Angular, where the magic prefix would be missing.

to

> However, that solution would be insufficient in case a library is built using partial compilation **while depending itself on** a VE-compiled library from earlier versions of Angular, where the magic prefix would be missing.

petebacondarwin commented on packages/compiler-cli/src/ngtsc/imports/test/emitter_spec.ts:
```diff
@@ -173,5 +209,111 @@ runInEachFileSystem(() => {
+      const decl = getDeclaration(program, _('/index.ts'), 'Foo', ts.isClassDeclaration);
+      const context = program.getSourceFile(_('/context.ts'))!;
+      const emitted = strategy.emit(new Reference(decl), context);
+      expect(emitted).not.toBeNull();
```
This line is redundant due to the next if statement.

petebacondarwin replied:
And also in the subsequent tests

petebacondarwin commented on packages/compiler-cli/test/ngtsc/incremental_semantic_changes_spec.ts:
```diff
@@ -1217,7 +1217,7 @@ runInEachFileSystem(() => {
             template: 'Dep',
           })
-          export class CmpDep {}
+          class CmpDep {}
```
Why these the changes in this file?

JoostK replied:
The test aims to verify that renaming an aliased export causes usage sites to be reemitted in incremental updates. The changes in this commit now cause the aliased exports to be ignored altogether, as there would have existed an export using the declared name. By removing the direct export we force the compiler to use the alias export, allowing the test to verify what it is intended to.

petebacondarwin reviewed (approved):
LGTM!

I wonder if we could cobble together an integration test that would directly show that the original issue is solved?

JoostK committed:
fixup! fix(compiler-cli): prefer non-aliased exports in reference emitters

JoostK commented:
> I wonder if we could cobble together an integration test that would directly show that the original issue is solved?

@petebacondarwin I added a compliance test that mimics the scenario, such that the golden files (and AOT output) capture that the correct export is chosen.

JoostK committed:
build: exclude compiler compliance tests from Saucelabs tests

The legacy-unit-tests-saucelabs job does not need to compile the
compiler compliance tests as they are not used in those tests. Since
the compliance tests can be configured to use specific compiler options,
the generic tsconfig that is used in legacy-unit-tests-saucelabs may not
succeed to compile the compliance tests so this commit excludes those
source files.

josephperrott reviewed (approved):
LGTM

Reviewed-for: dev-infra

mhevery reviewed (approved):
reviewed-for: global-approvers

mhevery commented:
[presubmit](http://test/OCL:371358029:BASE:371358052:1619803349883:b1f545ff)
